### ISSUE TITLE
Action result ProductSearchAutoComplete is copied from controller to controller

### ISSUE THREAD

**Issue by LubovKhokhlova** [nopSolutions/nopCommerce#3636]
Now the same code is used in ProductreViewController and in OrderController. It is necessary to put the same part in the method in the base controller, and this method is already called if necessary.

**Comment by dborovsky** [comment]
I would like implement this tasks.

**Comment by AndreiMaz** [comment]
@dborovsky thanks. please feel free to create a pull request

**Comment by dborovsky** [comment]
Hi Guys
I have questions.

I implemented as following:
**First way**

1. I created BaseAdminController constructor:
`public BaseAdminController(IHttpContextAccessor contextAccessor)
        {
            _contextAccessor = contextAccessor;
        }`
I need contextaccessor to get services

2. I have to call base constructor in derived controller like this:
`public ProductReviewController(CatalogSettings catalogSettings,
            ICustomerActivityService customerActivityService,
            IEventPublisher eventPublisher,
            IGenericAttributeService genericAttributeService,
            ILocalizationService localizationService,
            INotificationService notificationService,
            IPermissionService permissionService,
            IProductReviewModelFactory productReviewModelFactory,
            IProductService productService,
            IWorkContext workContext,
            IHttpContextAccessor contextAccessor,
            IWorkflowMessageService workflowMessageService) : base(contextAccessor)
        {
            ..........................................................
        }`

`
public IActionResult ProductSearchAutoComplete(string term)
        {
            var context = _contextAccessor.HttpContext;
            IWorkContext workContext = context.RequestServices.GetService<IWorkContext>();
            IPermissionService permissionService = context.RequestServices.GetService<IPermissionService>();
            IProductService productService = context.RequestServices.GetService<IProductService>();

........................................................................................................................
}
`
 It`s working, but every controller inherited from AdminbaseController must call base constructor with params. Is this good way?

**Second way**

We can initialize all services in AdminBaseController and in controllers too just calling base constructor.  

What do you think about it? Maybe do you have third way. Tell me, I`ll implement.

**Comment by davidandrewpowell** [comment]
I can see that the `ProductSearchAutoComplete(string)` method is used in the following admin controllers:

```
OrderController
ProductReviewController
ShoppingCartController
```

I don't think that the `BaseAdminController` is the correct place to move this method to.  Personally, I would reserve the base controller to house methods used by all (if not nearly all) inheriting controllers.  The `ProductSearchAutoComplete(string)` method does not seem to fit this criteria.

Moving the method to the base controller is going to involve changing all the controllers that don't even care about the autocomplete but now they are being forced to pass this dependency.  This also seems wrong.

If we did move the method to the base controller, then where does it stop?  Any time a controller method is called in multiple places should it be moved to the base controller?

Instead, I would suggest the method be moved to a new controller.  Maybe a `ProductSearchController` with a single action of `AutoComplete(string)` perhaps (the routing can be changed if it is necessary to keep the same URL).  It seems that a lot of controllers seem to house hundreds, if not thousands of lines of code with so many dependencies.  Personally, I would like to see a nice simple controller (think SOLID principles) with an action or actions only relating to product search keeping it very simple.

Therefore, with only a small number of dependencies on the controller it should also hopefully be straightforward to set up unit tests around this.

Please let me know your thoughts.

**Comment by dborovsky** [comment]
@davidandrewpowell Yea, I agree with you.  BaseController is no good place. I noticed too, that our controllers have many dependencies.

So I decided to ask

 @LubovKhokhlova  @AndreiMaz  @RomanovM  what do you think?

**Comment by DmitriyKulagin** [comment]
Closed #3636
